### ISSUE TITLE
IPv4 IPAddressString not containing another IPAddressString

### ISSUE THREAD

**Issue by everag** [seancfoley/IPAddress#26]
According to http://jodies.de/ipcalc?host=10.0.0.1&mask1=24&mask2= , `10.0.0.1/24` starts at `10.0.0.1` and ends at `10.0.0.254`. When trying to check with this lib if the upper bound of this subnet is contained within it, I'm getting `false`:

```
IPAddressString subnetStr = new IPAddressString("10.0.0.1/24");
IPAddressString ipStr = new IPAddressString("10.0.0.254");
subnetStr.contains(ipStr); // false
```

Oddly enough, both `subnetStr.getAddress().getLower()` and `subnetStr.getAddress().getUpper()` are both returning me `10.0.0.1`.

I'm not an expert on networking and IP addresses, so pardon me if I'm missing something obvious.

**Comment by seancfoley** [comment]
There are some common conventions with IP addresses.  One of them, as listed on the linked page, is to represent the subnet using the address with 0 host, and so when the prefix length is 24, the zero host is the address with 0 as the last segment.

This is indicated on the linked page where it says the network is 10.0.0.0/24

The default behaviour of this library follows that convention.  So to get the subnet you would do:

     IPAddressString subnetStr = new IPAddressString("10.0.0.0/24");

When you do not specify a zero host, it is just a single address, so the following is just the single address 10.0.0.1 with prefix length 24:

    IPAddressString subnetStr = new IPAddressString("10.0.0.1/24");

So that is why both lower and upper are 10.0.0.1 for that address.  You can call getCount() on the address to see it is just 1, a single address.

You can also get the subnet using toPrefixBlock() which converts the address to the subnet for the prefix length.

So your options are to use the subnet directly:

    IPAddressString subnetStr = new IPAddressString("10.0.0.0/24");
    IPAddressString ipStr = new IPAddressString("10.0.0.254");
    subnetStr.contains(ipStr); // true

or to get the subnet from the address:

    IPAddressString subnetStr = new IPAddressString("10.0.0.1/24");
    IPAddressString ipStr = new IPAddressString("10.0.0.254");
    subnetStr.getAddress().toPrefixBlock().contains(ipStr.getAddress()); // true

There is a third option where we do not use the convention first described in this answer (ie the zero host is the subnet), instead we interpret any address with a prefix length as the subnet, which is what you appear to be doing in the question.  That is not the default behaviour of the library.  To get that behaviour, instructions are [here](https://seancfoley.github.io/IPAddress/#_Toc525135538) under "Alternative options for Prefix Length Handling". The short of it is that you would call 

    AddressNetwork.setPrefixConfiguration(ALL_PREFIXED_ADDRESSES_ARE_SUBNETS);

However, I'd recommend that you just stick with the default behaviour and one of the first two options.

**Comment by everag** [comment]
Thanks for the quick response! I indeed noticed in that link that the network was `10.0.0.0` instead of `10.0.0.1`, but haven't realized that there was this standard for prefixed IPs. I tested your suggestion for using `toPrefixBlock()` and it works perfectly!

I think the main problem was that I am migrating from `commons-net` to `IPAddress` for IPv6 support, and apparently commons-net always considers prefixed IPs to be subnets. Since a lot of people will migrate to IPAddress due to that support (and also a WAY better API, really), maybe it's worth adding a migration guide doc. :-)

Thanks and keep up the good work!

**Comment by seancfoley** [comment]
If I remember correctly, commons-net has functions specifically for subnets and others for addresses, so that is one difference with this library. It can be very useful to represent both subnets and addresses with the same types, like in this library. When I created this library it was because libraries like that one did not meet my needs, the needs which I list in the goals section of the docs.  
Anyway, originally in earlier versions all prefixed addresses were subnets but I decided moving away from that provided more functionality, but I did keep that behavior as non-default behavior for backwards compatibility.
Thanks for the feedback, it's very useful to get feedback. I will keep your suggestion under consideration.
